src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,494 @@
+         * creates a new list if there is none yet.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(byte[] salt) {
+            Objects.requireNonNull(salt, "salt must not be null");
+            if (salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "Generic"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Returns a builder for building {@code Extract} and
+     * {@code ExtractThenExpand} objects.
+     *
+     * @return a new {@code Builder}
+     */
+    static Builder ofExtract() {
+        return new Builder().createBuilder();
+    }
+
+    /**
+     * Creates an {@code Expand} object.
+     *
+     * @param prk
+     *     the pseudorandom key; must not be {@code null}
+     * @param info
+     *     the optional context and application specific information (may be
+     *     {@code null}); the byte array is copied to prevent subsequent
+     *     modification
+     * @param length
+     *     the length of the output keying material (must be greater than 0)
+     *
+     * @implNote HKDF implementations will enforce that the length is less than
+     * 255 * HMAC length. Implementations will also enforce that the prk is at least as
+     * many bytes as the HMAC length.
+     *
+     * @return a new {@code Expand} object
+     *
+     * @throws NullPointerException
+     *     if {@code prk} is {@code null}
+     * @throws IllegalArgumentException
+     *     if {@code length} is not > 0
+     */
+    static Expand expandOnly(SecretKey prk, byte[] info, int length) {
+        if (prk == null) {
+            throw new NullPointerException("prk must not be null");
+        }
+        return new Expand(prk, info, length);
+    }
+
+    /**
+     * Defines the input parameters of an Extract operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Extract implements HKDFParameterSpec {
+
+        // HKDF-Extract(salt, IKM) -> PRK
+        private final List<SecretKey> ikms;
+        private final List<SecretKey> salts;
+
+        private Extract(List<SecretKey> ikms, List<SecretKey> salts) {
+            this.ikms = List.copyOf(ikms);
+            this.salts = List.copyOf(salts);
+        }
+
+        /**
+         * Returns an unmodifiable {@code List} of input keying material values
+         * in the order they were added. Returns an empty list if there are no
+         * input keying material values.
+         * <p>
+         * Input keying material values added by {@link Builder#addIKM(byte[])}
+         * are converted to a {@code SecretKeySpec} object.
```

Comment:
Do we need to say "empty arrays were discarded"? This is more precise and JCK might test on it.